worker/raft/raftbackstop/worker.go

```diff
@@ -0,0 +1,257 @@
+		return nil, errors.Annotate(err, "requesting current apiserver details")
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: func() error {
+			defer unsubscribe()
+			return w.loop()
+		},
+	}); err != nil {
+		unsubscribe()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+type Worker struct {
+	catacomb catacomb.Catacomb
+	config   Config
+
+	serverDetails chan apiserver.Details
+	configUpdated bool
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *Worker) loop() error {
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case details := <-w.serverDetails:
+			err := w.maybeRecoverCluster(details)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (w *Worker) maybeRecoverCluster(details apiserver.Details) error {
+	if w.configUpdated {
+		w.config.Logger.Infof("raft configuration already updated; waiting for raft worker restart")
+		return nil
+	}
+	if len(details.Servers) != 1 {
+		return nil
+	}
+	if _, found := details.Servers[w.config.Tag.Id()]; !found {
+		return nil
+	}
+	if w.config.Raft.State() == raft.Leader {
+		return nil
+	}
+
+	raftServers, err := w.getConfiguration()
+	if err != nil {
+		return errors.Annotate(err, "getting raft configuration")
+	}
+
+	numServers := len(raftServers)
+	localServer := raftServers[raft.ServerID(w.config.Tag.String())]
+	if localServer == nil {
+		return nil
+	}
+
+	localVote := localServer.Suffrage != raft.Nonvoter
+	if numServers == 1 && localVote {
+		// The server can vote and has quorum, so it can become leader.
+		return nil
+	}
+
+	err = w.recoverCluster(localServer)
+	return errors.Annotate(err, "recovering cluster")
+}
+
+func (w *Worker) recoverCluster(server *raft.Server) error {
+	// Make a configuration that can be written into the log, and append it.
+	newServer := *server
+	newServer.Suffrage = raft.Voter
+	configuration := raft.Configuration{
+		Servers: []raft.Server{newServer},
+	}
+	w.config.Logger.Infof("appending recovery configuration to log: %#v", configuration)
+	data, err := encodeConfiguration(configuration)
+
+	// Work out the last term and index.
+	lastIndex, err := w.config.LogStore.LastIndex()
+	if err != nil {
+		return errors.Annotate(err, "getting last log index")
+	}
+	var lastLog raft.Log
+	err = w.config.LogStore.GetLog(lastIndex, &lastLog)
+	if err != nil {
+		return errors.Annotate(err, "getting last log entry")
+	}
+
+	// Prepare log record to add.
+	record := raft.Log{
+		Index: lastIndex + 1,
+		Term:  lastLog.Term,
+		Type:  raft.LogConfiguration,
+		Data:  data,
+	}
+	if err := w.config.LogStore.StoreLog(&record); err != nil {
+		return errors.Annotate(err, "storing recovery configuration")
+	}
+	w.config.Logger.Infof("configuration appended")
```

Comment:
not sure we need this, maybe as debug only. log if there's an error but the append message above is sufficient?

Reply:
Yeah, I guess so - my thinking was that there's a message saying we're doing the recovery and then one saying it's done. Probably it's more interesting to me because I was debugging it, though. I'll turn it down to debug.

(Oh, I remember - I was getting a hang because of an error in syncLogStore! I can probably just lose it in that case.)

Reply:
Removed.